app/services/bulk_hosts_manager.rb

```diff
@@ -0,0 +1,35 @@
+class BulkHostsManager
+  def initialize(hosts:)
+    @hosts = hosts
+  end
+
+  def build(reboot: false)
+    # returns missed hosts
+    @hosts.select do |host|
+      success = true
+      begin
+        host.built(false) if host.build? && host.token_expired?
+        host.setBuild
+        host.power.reset if reboot && host.supports_power_and_running?
+      rescue => error
+        message = _('Failed to redeploy %s.') % host
+        Foreman::Logging.exception(message, error)
```

Comment:
We don't translate our logging messages. I see it already was done previously, but I'd suggest to improve it while you're at it